File: aicup/src/core/dokcon.js
const net = require('net');
const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');

const AI_BIN_NAME = './ai.bin';
const TMP_DIR = '.'///tmpfs'; // можно на /tmpfs
const SOCKET_PATH = process.env.SOCKET_PATH || '/tmp/dokcon.sock';

var log=function(...args){
  console.log.apply(console, [qap_time(), ...args]);
};

function qap_time() {
  const now = new Date();
  const utc = new Date(now.getTime() + now.getTimezoneOffset() * 60000);
  const moscowTime = new Date(utc.getTime() + 3 * 60 * 60000);
  const year = moscowTime.getFullYear();
  const month = String(moscowTime.getMonth() + 1).padStart(2, '0');
  const day = String(moscowTime.getDate()).padStart(2, '0');
  const hours = String(moscowTime.getHours()).padStart(2, '0');
  const minutes = String(moscowTime.getMinutes()).padStart(2, '0');
  const seconds = String(moscowTime.getSeconds()).padStart(2, '0');
  const milliseconds = String(moscowTime.getMilliseconds()).padStart(3, '0');
  return `${year}.${month}.${day} ${hours}:${minutes}:${seconds}.${milliseconds}`;
}
var func=x=>JSON.stringify(x.toString('binary'));
var emitter_on_data_decoder=(emitter,cb)=>{
  var rd=Buffer.from([]);
  emitter.on('data',data=>{
    //log('Received raw data, length=' + data.length+" value="+func(data.slice(0,128)));
    rd=Buffer.concat([rd,data]);
    for(;;){
      var e=rd.indexOf("\0");
      if(e<0)return log('e<0 e=',e);
      var en=e+1;
      var zpos=rd.indexOf('\0',en);
      if(zpos<0)return log('zpos<0 zpos=',e);
      var zn=zpos+1;
      var blen=rd.slice(0,e);
      var len=parseInt(blen.toString("binary"),10);
      if(rd.length<zn+len)return log('rd.length<zn+len ',{rdlen:rd.length,zn,len});
      var bz=rd.slice(en,en+zpos-en);var z=bz.toString("binary");
      var bmsg=rd.slice(zn,zn+len);var msg=bmsg.toString("binary");
      rd=rd.slice(zn+len);
      //log('rd=' + rd.length+" value="+func(rd.slice(0,128)));
      //log('z=' + z+" msg="+func(msg.slice(0,128)));
      cb(z,msg,bz,bmsg);
    }
  });
};

var stream_write_encoder=(stream,z)=>data=>{
  var sep=Buffer.from([0]);
  stream.write(Buffer.concat([
    Buffer.from(!data?"0":(data.length+""),"binary"),sep,
    Buffer.from(z,"binary"),sep,
    Buffer.from(data?data:"","binary")
  ]));
};

async function handleConnection(socket) {
  log('Client connected');
  let aiProcess = null;
  socket.setNoDelay(true);
  stream_write_encoder(socket,'hi from dokcon.js')('2025.10.18 12:01:08.493');
  //log('after hi');
  socket.on('close', () => {
    if (aiProcess) {
      log('Killing AI process due to socket close');
      aiProcess.kill();
      aiProcess = null;
    }
  });
  //log('a bit later');
  socket.on('error', err => {
    log('Socket error:', err);
    if (aiProcess) {
      aiProcess.kill();
      aiProcess = null;
    }
  });
  
  //log('a bit later 2');
  emitter_on_data_decoder(socket, async (z, msg, bz, bmsg) => {
    switch (z) {
      case 'ai_stdin':
          if (aiProcess) {
            // Пересылаем в stdin AI
            aiProcess.stdin.write(bmsg); // бинарно!
            //stream_write_encoder(socket, 'log')('ai_stdin.length='+bmsg.length);
            return;
          }else stream_write_encoder(socket, 'log')('i got ai_stdin but aiProcess is not started!');
          break;

      case 'ai_binary':
        log(`Saving binary of length ${bmsg.length}`);
        const filePath = path.join(TMP_DIR, AI_BIN_NAME);
        try {
          log('otpravlenZ');
          fs.writeFileSync(filePath, bmsg, { mode: 0o755 });
          log('otpravlen0');
          //stream_write_encoder(socket, 'log')('Binary saved');
          //log('otpravlen1');
          stream_write_encoder(socket, 'ai_binary_ack')(bmsg.length+"");/*
          log('otpravlen2');
          stream_write_encoder(socket, 'log')('ai_binary_ack otpravlen 1');
          log('otpravlen3');
          stream_write_encoder(socket, 'log')('ai_binary_ack otpravlen 2');
          log('otpravlen4');
          stream_write_encoder(socket, 'log')('ai_binary_ack otpravlen 3');
          log('otpravlen5');
          stream_write_encoder(socket, 'log')('ai_binary_ack otpravlen 4');
          log('otpravlen6');
          stream_write_encoder(socket, 'log')('ai_binary_ack otpravlen 5');
          log('otpravlen7');*/
        } catch (err) {
          log('FAILED TO WRITE BINARY:', err);
          stream_write_encoder(socket, 'log')(`Write error: ${err.message}`);
        }
        break;

      case 'ai_start':
        if (aiProcess) {
          stream_write_encoder(socket, 'log')('AI already running');
          break;
        }
        const aiBin = path.join(TMP_DIR, AI_BIN_NAME);
        if (!fs.existsSync(aiBin)) {
          stream_write_encoder(socket, 'log')('Binary not found');
          break;
        }
        log('Starting AI process');
        log(aiBin);
        aiProcess = spawn("./"+aiBin,[],{stdio:['pipe','pipe','pipe'],windowsHide:true});

        // Перенаправляем stdout/stderr через стрим-протокол
        aiProcess.stdout.on('data', stream_write_encoder(socket, 'ai_stdout'));
        aiProcess.stderr.on('data', stream_write_encoder(socket, 'ai_stderr'));

        aiProcess.on('close', (code) => {
          stream_write_encoder(socket, 'log')(`AI exited with code ${code}`);
          aiProcess = null;
        });
        break;

      case 'ai_stop':
        if (aiProcess) {
          aiProcess.kill();
          stream_write_encoder(socket, 'log')('AI killed');
        } else {
          stream_write_encoder(socket, 'log')('No AI process running');
        }
        break;

      default:
        stream_write_encoder(socket, 'log')(`Unknown channel: ${z}`);
        log(`Unknown channel: ${z}`);
    }
  });
}

function startServer(useUnixSocket) {
  if (useUnixSocket) {
    if (fs.existsSync(SOCKET_PATH)) fs.unlinkSync(SOCKET_PATH);
    const server = net.createServer(handleConnection);
    server.listen(SOCKET_PATH, () => {
      log(`Server listening on Unix socket ${SOCKET_PATH}`);
    });
    fs.chmodSync(SOCKET_PATH, 0o777);
  } else {
    const PORT = 4000;
    const server = net.createServer(handleConnection);
    server.listen(PORT, () => {
      log(`Server listening on TCP port ${PORT}`);
    });
  }
}

// --- Клиент ---

let lastErrNum = -1;
const c_out = [];
const c_err_lines = [];
let err_buffer = ''; // для незавершённых строк

async function runClient(host, port, binaryFilePath) {
  return new Promise((resolve, reject) => {
    const client = new net.Socket();
    client.connect(port, host, async () => {
      console.log('Connected to server');

      // Читаем бинарник
      const binData = fs.readFileSync(binaryFilePath);

      // Отправляем бинарь
      stream_write_encoder(client, 'ai_binary')(binData);
      console.log('Sent binary');

      // Ждём подтверждение
      emitter_on_data_decoder(client, (z, msg) => {
        if (z === 'log' && msg.includes('Binary saved')) {
          // Запускаем ИИ
          stream_write_encoder(client, 'ai_start')('');
          console.log('Sent start command');
        }

        if (z === 'ai_stdout') {
          console.log('AI stdout:', msg);
          c_out.push(msg);
        }


        if (z === 'ai_stderr') {
          log('AI stderr:', JSON.stringify(msg)); // видим сырые фрагменты
          err_buffer += msg;

          // Разбиваем по \n (т.к. endl → \r\n, но \r не мешает)
          const lines = err_buffer.split('\n');
          err_buffer = lines.pop(); // последний — незавершённый

          for (const line of lines) {
            if (!line.trim()) continue;

            if (line.startsWith('ERR')) {
              const numStr = line.slice(3);
              const num = parseInt(numStr, 10);

              if (!isNaN(num)) {
                if (lastErrNum >= 0 && num !== lastErrNum + 1) {
                  console.log(`fail!!! Expected ERR${lastErrNum + 1}, got ERR${num} (line: ${line})`);
                }
                lastErrNum = num;
              } else {
                console.log(`fail!!! ERR followed by non-number: "${numStr}" (line: ${line})`);
              }
            }
            // Можно логировать полную строку
            console.log('Full line:', line);
          }
        }

        if (z === 'log') {
          console.log('Log:', msg);
          if (msg.includes('exited')) {
            client.destroy();
          }
        }
      });

      client.on('close', () => {
        console.log('Connection closed');
        resolve();
      });

      client.on('error', err => {
        reject(err);
      });
    });
  });
}

async function runClientUnix(socketPath, binaryFilePath) {
  return new Promise((resolve, reject) => {
    const client = net.createConnection(socketPath);

    client.on('connect', async () => {
      console.log('Connected to Unix socket:', socketPath);

      // Читаем бинарник
      const binData = fs.readFileSync(binaryFilePath);

      // Отправляем бинарь
      stream_write_encoder(client, 'ai_binary')(binData);
      console.log('Sent binary');

      let binaryAckReceived = false;

      emitter_on_data_decoder(client, (z, msg, bz, bmsg) => {
        if (z === 'log') {
          console.log('Log:', msg);
        }

        if (z === 'ai_binary_ack' && !binaryAckReceived) {
          binaryAckReceived = true;
          console.log('Received ai_binary_ack, starting AI...');
          stream_write_encoder(client, 'ai_start')('hm...');
        }

        if (z === 'ai_stdout') {
          console.log('AI stdout:', msg);
        }

        if (z === 'ai_stderr') {
          log('AI stderr:', msg);
        }

        // Опционально: завершить, если AI вышел
        if (z === 'log' && msg.includes('exited')) {
          client.end();
        }
        log("msg from z='"+z+"'");
      });

      client.on('close', () => {
        console.log('Connection closed');
        resolve();
      });

      client.on('error', (err) => {
        log('Socket error:', err.message);
        reject(err);
      });
    });

    client.on('error', (err) => {
      log('Failed to connect to Unix socket:', err.message);
      reject(err);
    });
  });
}

// --- Запуск ---

if (process.argv[2] === 'server_unix') {
  startServer(true);
} else if (process.argv[2] === 'server_tcp') {
  startServer(false);
} else if (process.argv[2] === 'client_tcp' && process.argv[3] && process.argv[4] && process.argv[5]) {
  const host = process.argv[3];
  const port = parseInt(process.argv[4]);
  const file = process.argv[5];
  runClient(host, port, file).catch(log);
} else if (process.argv[2] === 'client_unix' && process.argv[3] && process.argv[4]) {
  const socketPath = process.argv[3];
  const binaryFilePath = process.argv[4];
  runClientUnix(socketPath, binaryFilePath).catch(log);
} else {
  console.log('Usage: node dokcon.js [server_unix|server_tcp|client_tcp <host> <port> <file>]');
  //node dokcon.js client_tcp 127.0.0.1 4000 ai.exe
}

File: aicup/src/core/dokcon.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <iomanip>
#include <chrono>
#include <functional>
#include <thread>
#include <atomic>
#include <mutex>
#include <vector>
#include <cstring>
#include <cstdlib>
#include <cstdio>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <signal.h>
#include <sys/wait.h>
#include <fstream>
#include <memory>
using namespace std;

std::time_t timegm_portable(std::tm* tm) {
#if defined(_WIN32) || defined(_WIN64)
  return _mkgmtime(tm);
#else
  return timegm(tm);
#endif
}

void add_hours_to_tm(std::tm& tm, int hours) {
  std::time_t t = timegm_portable(&tm);
  t += static_cast<std::time_t>(hours) * 3600; // сдвигаем UTC-время
#if defined(_WIN32) || defined(_WIN64)
  gmtime_s(&tm, &t);
#else
  gmtime_r(&t, &tm);
#endif
}

string qap_time() {
  using namespace std::chrono;
  system_clock::time_point now = system_clock::now();
  auto duration = now.time_since_epoch();
  auto sec = duration_cast<seconds>(duration);
  auto millis = duration_cast<milliseconds>(duration - sec);
  std::time_t t_c = sec.count();

  std::tm utc_tm;
#if defined(_WIN32) || defined(_WIN64)
  gmtime_s(&utc_tm, &t_c);
#else
  gmtime_r(&t_c, &utc_tm);
#endif

  // Москва = UTC + 3
  add_hours_to_tm(utc_tm, 3);

  std::ostringstream oss;
  oss << std::put_time(&utc_tm, "%Y.%m.%d %H:%M:%S")
      << '.' << std::setfill('0') << std::setw(3) << millis.count();
  return oss.str();
}

string qap_zchan_write(const string&z,const string&data){
  string n=std::to_string(data.size());auto sep=string("\0",1);
  return n+sep+z+sep+data;
}

struct emitter_on_data_decoder{
  typedef std::function<void(const string&, const string&)> t_cb;
  t_cb cb;
  string buffer;
  struct t_parse_result{
    unsigned char s='o';
    string to_str()const{
      if(s=='o')return "WTF?";
      if(s=='0')return "wait_size";
      if(s=='1')return "wait_zchan";
      if(s=='2')return "wait_data";
      if(s=='s')return "size_atk";
      if(s=='d')return "digit_atk";
      if(s=='z')return "zchan_atk";
      if(s=='3')return "ok?";
      return "no_impl";
    }
    bool ok(){switch(s){case '0':case '1':case '2':return true;}return false;}
  };
  static bool is_digits(const string&s){for(char c:s)if(c<'0'||c>'9')return false;return true;}
  t_parse_result feed(const char* data, size_t len) {
    buffer.append(data, len);
    while (true) {
      auto e1 = buffer.find('\0');
      if (e1 == string::npos)return {'0'};
      string len_str = buffer.substr(0, e1);
      if(len_str.size()>=7)return {'s'};
      if(!is_digits(len_str))return {'d'};
      int len = stoi(len_str);
      auto e2 = buffer.find('\0', e1 + 1);
      if (e2 == string::npos)return {'1'};
      if(e2>512)return {'z'};
      int total = e2 + 1 + len;
      if (buffer.size() < total)return {'2'};
      string z = buffer.substr(e1 + 1, e2 - e1 - 1);
      string payload = buffer.substr(e2 + 1, len);
      buffer.erase(0, total);
      cb(z, payload);
    }
    return {'3'};
  }
};

// Unix Socket Server
class UnixSocketServer {
private:
  std::string socket_path_;
  int server_fd_ = -1;
  std::atomic<bool> running_{false};
  std::thread server_thread_;
  
  std::function<void(int client_fd)> client_handler_;

public:
  UnixSocketServer(const std::string& socket_path, 
          std::function<void(int client_fd)> client_handler)
    : socket_path_(socket_path), client_handler_(client_handler) {}

  ~UnixSocketServer() {
    stop();
  }

  bool start() {
    // Удаляем старый socket файл если существует
    ::unlink(socket_path_.c_str());

    server_fd_ = socket(AF_UNIX, SOCK_STREAM, 0);
    if (server_fd_ == -1) {
      std::cerr << qap_time() << " Failed to create socket" << std::endl;
      return false;
    }

    struct sockaddr_un addr;
    std::memset(&addr, 0, sizeof(addr));
    addr.sun_family = AF_UNIX;
    std::strncpy(addr.sun_path, socket_path_.c_str(), sizeof(addr.sun_path) - 1);

    if (bind(server_fd_, (struct sockaddr*)&addr, sizeof(addr)) == -1) {
      std::cerr << qap_time() << " Bind failed" << std::endl;
      close(server_fd_);
      return false;
    }

    if (listen(server_fd_, 10) == -1) {
      std::cerr << qap_time() << " Listen failed" << std::endl;
      close(server_fd_);
      return false;
    }

    // Устанавливаем права на socket
    chmod(socket_path_.c_str(), 0777);

    running_ = true;
    server_thread_ = std::thread([this]() { run(); });

    std::cout << qap_time() << " Server listening on Unix socket " << socket_path_ << std::endl;
    return true;
  }

  void stop() {
    running_ = false;
    if (server_fd_ != -1) {
      close(server_fd_);
      server_fd_ = -1;
    }
    if (server_thread_.joinable()) {
      server_thread_.join();
    }
    ::unlink(socket_path_.c_str());
  }

private:
  void run() {
    while (running_) {
      int client_fd = accept(server_fd_, nullptr, nullptr);
      if (client_fd == -1) {
        if (running_) {
          std::cerr << qap_time() << " Accept failed" << std::endl;
        }
        continue;
      }

      // Обрабатываем клиента в отдельном потоке
      std::thread client_thread([this, client_fd]() {
        client_handler_(client_fd);
        close(client_fd);
      });
      client_thread.detach();
    }
  }
};

// AI Process Manager
class AIProcessManager {
private:
  pid_t ai_pid_ = -1;
  int stdin_pipe_[2] = {-1, -1};
  int stdout_pipe_[2] = {-1, -1};
  int stderr_pipe_[2] = {-1, -1};
  
  std::function<void(const std::string&, const std::string&)> send_message_;
  std::string binary_path_;
  
  std::thread stdout_thread_;
  std::thread stderr_thread_;
  std::atomic<bool> running_{false};

public:
  AIProcessManager(const std::string& binary_path, 
          std::function<void(const std::string&, const std::string&)> send_message)
    : binary_path_(binary_path), send_message_(send_message) {}

  ~AIProcessManager() {
    stop();
  }

  bool start() {
    if (ai_pid_ != -1) {
      return false; // Уже запущен
    }

    // Создаем пайпы
    if (pipe(stdin_pipe_) == -1 || pipe(stdout_pipe_) == -1 || pipe(stderr_pipe_) == -1) {
      return false;
    }

    ai_pid_ = fork();
    if (ai_pid_ == -1) {
      close_pipes();
      return false;
    }

    if (ai_pid_ == 0) { // Child process
      close(stdin_pipe_[1]); // Close write end of stdin
      close(stdout_pipe_[0]); // Close read end of stdout
      close(stderr_pipe_[0]); // Close read end of stderr

      // Перенаправляем стандартные потоки
      dup2(stdin_pipe_[0], STDIN_FILENO);
      dup2(stdout_pipe_[1], STDOUT_FILENO);
      dup2(stderr_pipe_[1], STDERR_FILENO);

      // Запускаем AI бинарник
      execl(binary_path_.c_str(), binary_path_.c_str(), nullptr);
      
      // Если дошли сюда - ошибка
      exit(1);
    } else { // Parent process
      close(stdin_pipe_[0]); // Close read end of stdin
      close(stdout_pipe_[1]); // Close write end of stdout
      close(stderr_pipe_[1]); // Close write end of stderr

      running_ = true;
      
      // Запускаем потоки для чтения stdout/stderr
      stdout_thread_ = std::thread([this]() { read_loop(stdout_pipe_[0], "ai_stdout"); });
      stderr_thread_ = std::thread([this]() { read_loop(stderr_pipe_[0], "ai_stderr"); });

      send_message_("log", "AI process started");
      return true;
    }
  }

  void stop() {
    running_ = false;
    
    if (ai_pid_ != -1) {
      kill(ai_pid_, SIGTERM);
      waitpid(ai_pid_, nullptr, 0);
      ai_pid_ = -1;
    }

    close_pipes();

    if (stdout_thread_.joinable()) stdout_thread_.join();
    if (stderr_thread_.joinable()) stderr_thread_.join();
  }

  bool write_to_stdin(const std::string& data) {
    if (stdin_pipe_[1] == -1) return false;
    
    ssize_t written = write(stdin_pipe_[1], data.data(), data.size());
    return written == static_cast<ssize_t>(data.size());
  }

  bool is_running() const { return ai_pid_ != -1; }

private:
  void close_pipes() {
    if (stdin_pipe_[0] != -1) { close(stdin_pipe_[0]); stdin_pipe_[0] = -1; }
    if (stdin_pipe_[1] != -1) { close(stdin_pipe_[1]); stdin_pipe_[1] = -1; }
    if (stdout_pipe_[0] != -1) { close(stdout_pipe_[0]); stdout_pipe_[0] = -1; }
    if (stdout_pipe_[1] != -1) { close(stdout_pipe_[1]); stdout_pipe_[1] = -1; }
    if (stderr_pipe_[0] != -1) { close(stderr_pipe_[0]); stderr_pipe_[0] = -1; }
    if (stderr_pipe_[1] != -1) { close(stderr_pipe_[1]); stderr_pipe_[1] = -1; }
  }

  void read_loop(int pipe_fd, const std::string& channel) {
    char buffer[4096];
    while (running_) {
      ssize_t n = read(pipe_fd, buffer, sizeof(buffer));
      if (n <= 0) break;
      
      send_message_(channel, std::string(buffer, n));
    }
  }
};

// Client Session Handler
class ClientSession {
private:
  int client_fd_;
  std::atomic<bool> connected_{true};
  emitter_on_data_decoder decoder_;
  AIProcessManager ai_manager_;
  
  std::string tmp_dir_ = ".";
  std::string ai_bin_name_ = "./ai.bin";
  
  std::mutex write_mutex_;

public:
  ClientSession(int client_fd, const std::string& tmp_dir = ".")
    : client_fd_(client_fd), tmp_dir_(tmp_dir),
     ai_manager_(tmp_dir + "/" + ai_bin_name_, 
           [this](const std::string& z, const std::string& data) {
             send_message(z, data);
           }) {
    
    decoder_.cb = [this](const std::string& z, const std::string& payload) {
      handle_message(z, payload);
    };
    
    // Отправляем приветствие
    send_message("hi from dokcon.cpp", "2025.10.18 12:01:08.493");
    
    // Запускаем цикл чтения
    std::thread([this]() { read_loop(); }).detach();
  }

  ~ClientSession() {
    connected_ = false;
    close(client_fd_);
  }

private:
  void read_loop() {
    char buffer[4096];
    while (connected_) {
      ssize_t n = read(client_fd_, buffer, sizeof(buffer));
      if (n <= 0) {
        connected_ = false;
        break;
      }
      decoder_.feed(buffer, n);
    }
  }

  void send_message(const std::string& z, const std::string& data) {
    std::lock_guard<std::mutex> lock(write_mutex_);
    std::string frame = qap_zchan_write(z, data);
    ssize_t sent = write(client_fd_, frame.data(), frame.size());
    if (sent != static_cast<ssize_t>(frame.size())) {
      connected_ = false;
    }
  }

  void handle_message(const std::string& z, const std::string& msg) {
    if (z == "ai_stdin") {
      if (ai_manager_.is_running()) {
        ai_manager_.write_to_stdin(msg);
      } else {
        send_message("log", "i got ai_stdin but aiProcess is not started!");
      }
    } else if (z == "ai_binary") {
      std::string file_path = tmp_dir_ + "/" + ai_bin_name_;
      std::ofstream file(file_path, std::ios::binary);
      if (file) {
        file.write(msg.data(), msg.size());
        file.close();
        chmod(file_path.c_str(), 0755);
        send_message("ai_binary_ack", std::to_string(msg.size()));
      } else {
        send_message("log", "Failed to write binary file");
      }
    } else if (z == "ai_start") {
      if (ai_manager_.is_running()) {
        send_message("log", "AI already running");
      } else {
        if (ai_manager_.start()) {
          send_message("log", "AI process started successfully");
        } else {
          send_message("log", "Failed to start AI process");
        }
      }
    } else if (z == "ai_stop") {
      if (ai_manager_.is_running()) {
        ai_manager_.stop();
        send_message("log", "AI killed");
      } else {
        send_message("log", "No AI process running");
      }
    } else {
      send_message("log", "Unknown channel: " + z);
    }
  }
};

// Main application
class DokconServer {
public:
  std::unique_ptr<UnixSocketServer> server_;
  std::string socket_path_ = "/tmp/dokcon.sock";
  std::string tmp_dir_ = ".";

public:
  void run() {
    std::cout << qap_time() << " Starting dokcon.cpp server..." << std::endl;
    
    server_ = std::make_unique<UnixSocketServer>(
      socket_path_, 
      [this](int client_fd) { 
        handle_client(client_fd); 
      }
    );
    
    if (!server_->start()) {
      std::cerr << qap_time() << " Failed to start server" << std::endl;
      return;
    }
    
    // Ждем сигнала завершения
    std::cout << "Press Enter to stop..." << std::endl;
    std::cin.get();
    
    server_->stop();
  }

private:
  void handle_client(int client_fd) {
    std::cout << qap_time() << " Client connected" << std::endl;
    
    // Создаем сессию для клиента (управляется внутри класса)
    ClientSession session(client_fd, tmp_dir_);
    
    // Ждем пока клиент отключится
    while (true) {
      std::this_thread::sleep_for(std::chrono::seconds(1));
      // Проверка соединения происходит внутри ClientSession
    }
    
    std::cout << qap_time() << " Client disconnected" << std::endl;
  }
};

int main(int argc, char* argv[]) {
  if (argc > 1 && std::string(argv[1]) == "server_unix") {
    DokconServer server;
    server.socket_path_=std::getenv("SOCKET_PATH");
    server.run();
  } else {
    std::cout << "Usage: " << argv[0] << " [server_unix]" << std::endl;
  }
  
  return 0;
}